### PR TITLE
LedgerConsensus cleanup

### PR DESCRIPTION
**Pull request by nbougalis**
As part of our ongoing code-maintenance, I did the following cleanups in `LedgerConsensus`, `LedgerProposal` and `NetworkOPsImp`.

I made it a point to comment and add exposition to help future readers of the code, but tried to make minimal logic changes (and to isolate them in their own commit).

For your reviewing pleasure @seelabs, @HowardHinnant, @rec. Also, @JoelKatz since this touches critical consensus code.

### CODE REVIEW

**Comment by seelabs** [comment]
:+1: Nice changes.

**Comment by HowardHinnant** [comment]
Looks like a missing include here:
clang++ -o build/clang.debug.nounity/src/ripple/app/ledger/LedgerConsensus.o -c -Wall -frtti -std=c++11 -Wno-invalid-offsetof -Wno-mismatched-tags -Wno-deprecated-register -Wno-sign-compare -Wno-char-subscripts -Wno-format -g -Wno-redeclared-class-member -Wno-deprecated -Wno-deprecated-declarations -Wno-unused-variable -Wno-unused-function -DOPENSSL_NO_SSL2 -DDEPRECATED_IN_MAC_OS_X_VERSION_10_7_AND_LATER -DHAVE_USLEEP=1 -DDEBUG -D_DEBUG -DBOOST_ASIO_HAS_STD_ARRAY -D_FILE_OFFSET_BITS=64 -DBEAST_COMPILE_OBJECTIVE_CPP=1 -I/usr/local/Cellar/openssl/1.0.1f/include -I/Users/howardhinnant/Development/boost_1_57_0 -Isrc -Isrc/beast -Ibuild/proto -Isrc/soci/src src/ripple/app/ledger/LedgerConsensus.cpp
In file included from src/ripple/app/ledger/LedgerConsensus.cpp:22:
src/ripple/app/ledger/LedgerConsensus.h:71:5: error: unknown type name 'InboundTransactions'
    InboundTransactions& inboundTransactions, LocalTxs& localtx,
    ^

**Comment by JoelKatz** [comment]
:+1: All I found were a few nits.

**Comment by nbougalis** [comment]
Thanks @HowardHinnant. I missed that; will fix.

**Comment by JoelKatz** [comment]
I had a +1 here, but I temporarily removed it. I have to check how you check the signature without the public key stored in the proposal. I don't see how you could do it.

**Comment by nbougalis** [comment]
The public key is still in the proposal and is used. But yes, by all means do verify!

File: src/ripple/app/ledger/LedgerConsensus.cpp [hunk 1]
```diff
@@ -66,6 +66,22 @@ class LedgerConsensusImp
...
     , public CountedObject <LedgerConsensusImp>
 {
+private:
+    enum class State
+    {
+        // We haven't closed our ledger yet, but others might have
+        open,
+
+        // Establishing consensus
+        establish,
```
Comment (rec) [inline src/ripple/app/ledger/LedgerConsensus.cpp]: This one word stands out as a verb in the infinitive where the others act like adjectives...?

Why not call it `establishing`?

File: src/ripple/app/ledger/LedgerConsensus.cpp [hunk 2]
```diff
@@ -196,21 +211,21 @@ class LedgerConsensusImp
         else
             ret["synched"] = false;
 
-        switch (mState)
+        switch (state_)
         {
-        case lcsPRE_CLOSE:
```
Comment (rec) [inline src/ripple/app/ledger/LedgerConsensus.cpp]: SO much better!

File: src/ripple/app/ledger/LedgerConsensus.cpp [hunk 3]
```diff
@@ -556,7 +572,8 @@ class LedgerConsensusImp
         mPreviousLedger = newLCL;
         mPrevLedgerHash = lclHash;
 
-        WriteLog (lsINFO, LedgerConsensus) << "Have the consensus ledger " << mPrevLedgerHash;
+        WriteLog (lsINFO, LedgerConsensus) <<
+            "Have the consensus ledger " << mPrevLedgerHash;
```
Comment (rec) [inline src/ripple/app/ledger/LedgerConsensus.cpp]: Why use one line where two will do?  :-P

File: src/ripple/protocol/impl/STValidation.cpp [hunk 4]
```diff
@@ -57,20 +57,16 @@ STValidation::STValidation (
...
-    sign (signingHash, raPriv);
-}
-
-void STValidation::sign (uint256& signingHash, RippleAddress const& raPriv)
+uint256 STValidation::sign (RippleAddress const& raPriv)
 {
     setFlag (vfFullyCanonicalSig);
 
-    signingHash = getSigningHash ();
+    uint256 signingHash = getSigningHash ();
```
Comment (rec) [inline src/ripple/protocol/impl/STValidation.cpp]: Could be `auto`.

**Comment by rec** [comment]
:+1: from me - tiny quibbles you can address or not.

**Comment by JoelKatz** [comment]
I spoke to @nbougalis and I think we should just rip out the support for proposals without a previous ledger hash. They haven't been sent in at least a year and the code to support them is confusing and hard to understand.

**Comment by vinniefalco** [comment]
@JoelKatz @nbougalis It would be great to require that proposals and validations include the previous ledger hash, doing so would simplify some distributed computation problems.

**Comment by HowardHinnant** [comment]
Looks good to me.  Holding off on the thumbs up until I can verify the missing include is fixed.

**Comment by HowardHinnant** [comment]
Oops!  I've got a repeatable/reliable crash a few seconds after startup.  It has happened 3 out 3 times so far.  Stack dump:  http://codepad.org/btDAJr37

**Comment by seelabs** [comment]
Nice find @HowardHinnant: One thought: When `shared_ptr<LedgerConsensusImp>` was converted to a `unique_ptr` we caused a problem in `shared_from_this` which requires at least one shared pointer of that object to exist.

**Comment by HowardHinnant** [comment]
I just changed that back to `shared_ptr` and I still got the identical crash.

**Comment by nbougalis** [comment]
Oops! Will troubleshoot that - thanks Howard.

**Comment by HowardHinnant** [comment]
I think this is the line that's throwing: https://github.com/nbougalis/rippled/blob/b8b454342dc678122f1ee8e09c24288213110a6a/src/ripple/app/ledger/LedgerConsensus.cpp#L1732  (from `shared_from_this()`).

**Comment by HowardHinnant** [comment]
Ok, I resolved the crash by reverting the entire commit: https://github.com/nbougalis/rippled/commit/b8b454342dc678122f1ee8e09c24288213110a6a

I had first only tried to revert the type of `mConsensus`, but that wasn't enough.  That is subtle!!!  There are some really hairy ownership issues going on here.

**Comment by JoelKatz** [comment]
When you use shared_ptr's to manage the lifetime of an object, it changes a lot of things about that object. In this case, the failure to remove the inheritance from std::enabled_shared_from_this caused this error to be visible only at run time.

**Comment by seelabs** [comment]
:+1:

**Comment by nbougalis** [comment]
It seems that all the reviewers have thumbed this up. I'm going to mark it passed.

**Pull request closed without merging**